### PR TITLE
Prevent infinite loops in wasm hot reload

### LINKED ISSUE #33209: dotnet-watch: WebAssembly check that verifies if delta is applied can spin indefinitely
The VerifyDeltaApplied check spins indefinitely if a browser does not establish a websocket connection back to the app. This also means the delta failed to apply which should cause the app to be treated as a rude edit.

### PR DESCRIPTION
**Pull request by pranavkm**
* If the WebSocket connection to the browser failed to be established,
dotnet-watch will loop indefinitely as WaitAsync never gets scheduled. This change
adds a limit to how long the app waits, and also terminates early if no client is currently connected

* In addition, it's possible for the TLS connection to fail to be established even though the dev-certs is installed.
This is particularly apparent when developing in WSL. In such an event, attempt to fallback to the non-tls WS endpoint.

Fixes https://github.com/dotnet/aspnetcore/issues/33209

### CODE REVIEW

File: src/BuiltInTools/dotnet-watch/Filters/BrowserRefreshServer.cs [hunk 1]
```diff
@@ -38,18 +38,25 @@ public BrowserRefreshServer(IReporter reporter)
...
 
-            var useTls = await ShouldUseHttps();
+            var supportsTLS = await SupportsTLS();
 
             _refreshServer = new HostBuilder()
                 .ConfigureWebHost(builder =>
                 {
                     builder.UseKestrel();
-                    builder.UseUrls(useTls ? $"https://{hostName}:0" : $"http://{hostName}:0");
+                    if (supportsTLS)
```
Comment (captainsafia) [inline src/BuiltInTools/dotnet-watch/Filters/BrowserRefreshServer.cs:52]: Do we need to bother with this check if we can fallback on the logic that makes the WebSocket connection to the valid URL on the client-side?

Feels like we can avoid making the additional process call here.

File: src/BuiltInTools/dotnet-watch/Filters/BrowserRefreshServer.cs [hunk 2]
```diff
@@ -61,23 +68,26 @@ public async ValueTask<string> StartAsync(CancellationToken cancellationToken)
...
+                    .Replace("https://127.0.0.1", "wss://localhost", StringComparison.Ordinal));
             }
 
-            return serverUrl
-                .Replace("https://", "wss://", StringComparison.Ordinal)
-                .Replace("http://", "ws://", StringComparison.Ordinal);
+            return new[]
+            {
+                serverUrls
+                    .First()
```
Comment (captainsafia) [inline src/BuiltInTools/dotnet-watch/Filters/BrowserRefreshServer.cs:87]: Why do we only pick the first URL here?

File: src/BuiltInTools/dotnet-watch/HotReload/BlazorWebAssemblyDeltaApplier.cs [hunk 3]
```diff
@@ -72,43 +72,43 @@ public async ValueTask ReportDiagnosticsAsync(DotNetWatchContext context, IEnume
...
-                return IsDeltaApplied(result);
+                    var result = await context.BrowserRefreshServer.ReceiveAsync(_receiveBuffer, cancellationToken);
+                    if (result is null)
+                    {
+                        // A null result indicates no clients are connected. No deltas could have been applied in this state.
+                        _reporter.Verbose("No client is connected to ack deltas");
+                        return false;
+                    }
+
+                    if (IsDeltaApplied(result.Value))
```
Comment (BrennanConroy) [inline src/BuiltInTools/dotnet-watch/HotReload/BlazorWebAssemblyDeltaApplier.cs:92]: nit: return IsDeltaApplied

File: src/BuiltInTools/BrowserRefresh/WebSocketScriptInjection.js [hunk 4]
```diff
@@ -159,4 +166,27 @@ setTimeout(function () {
...
+        clearEventListeners();
+        reject(error);
+      }
+
+      function clearEventListeners() {
+        webSocket.removeEventListener('open', onOpen);
+        webSocket.removeEventListener('error', onError);
+      }
+
+      webSocket.addEventListener('open', onOpen);
```
[the remark this answers is not in the document]
Comment (pranavkm) [reply src/BuiltInTools/BrowserRefresh/WebSocketScriptInjection.js:188]: We shouldn't get to line44 until `getWebSocket` is awaited before which point the event listener is removed.

File: src/BuiltInTools/dotnet-watch/Filters/BrowserRefreshServer.cs (hunk 1 above)
Comment (pranavkm) [reply src/BuiltInTools/dotnet-watch/Filters/BrowserRefreshServer.cs:52]: Kestrel does not like it if you ask it to bind to a TLS port without the certs installed. So we do need to do some due-diligence here.

File: src/BuiltInTools/dotnet-watch/Filters/BrowserRefreshServer.cs (hunk 2 above)
Comment (pranavkm) [reply src/BuiltInTools/dotnet-watch/Filters/BrowserRefreshServer.cs:87]: Users can only provide one URL via an environment variable. We previously had this .First https://github.com/dotnet/sdk/pull/18039/files#diff-2696395053559822320fce8c67b5b71b00f693255b29906d02c6427be3e9533fL69 here.

File: src/BuiltInTools/dotnet-watch/HotReload/BlazorWebAssemblyDeltaApplier.cs (hunk 3 above)
Comment (pranavkm) [reply src/BuiltInTools/dotnet-watch/HotReload/BlazorWebAssemblyDeltaApplier.cs:92]: It's inside a loop though. We want it to return true the first time it succeeds.

Comment (BrennanConroy) [reply src/BuiltInTools/dotnet-watch/HotReload/BlazorWebAssemblyDeltaApplier.cs:92]: Ah didn't notice it was in the loop

File: src/BuiltInTools/BrowserRefresh/WebSocketScriptInjection.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/BuiltInTools/BrowserRefresh/WebSocketScriptInjection.js:188]: Don't use the error event prefer close. The docs aren't great for it, but through experience we've seen that close will always get called whereas error isn't. And if error is called, closed will be called.

You can see in SignalR that we just log in the error handler:
https://github.com/dotnet/aspnetcore/blob/e0139a1f55fa7eb7b7d700e27aa0f9d1a5720041/src/SignalR/clients/ts/signalr/src/WebSocketTransport.ts#L89-L97

[the remark this answers is not in the document]
Comment (pranavkm) [reply src/BuiltInTools/BrowserRefresh/WebSocketScriptInjection.js:188]: Aren't you using the error event in the code that you linked to? Is this what I should be using as reference: https://github.com/dotnet/aspnetcore/blob/e0139a1f55fa7eb7b7d700e27aa0f9d1a5720041/src/SignalR/clients/ts/signalr/src/WebSocketTransport.ts#L112-L130?

[the remark this answers is not in the document]
Comment (BrennanConroy) [reply src/BuiltInTools/BrowserRefresh/WebSocketScriptInjection.js:188]: > Aren't you using the error event in the code that you linked to?

Logging only.

> Is this what I should be using as reference

Yeah

[the remark this answers is not in the document]
Comment (pranavkm) [reply src/BuiltInTools/BrowserRefresh/WebSocketScriptInjection.js:203]: Updated

**Review by BrennanConroy — APPROVED** [review]

**Review by captainsafia — APPROVED** [review]

**Review by captainsafia — APPROVED** [review]

**Pull request merged**